calyx-py/calyx/builder_util.py

```diff
@@ -0,0 +1,190 @@
+    1. Within {comp}, creates a combinational group called {cell}_group.
+    2. Within the group, creates a {cell} that computes sums of {width}.
+    3. Puts the values {a} and {b} into {cell}.
+    4. Returns the summing cell and the overall group.
+    """
+    add_cell = comp.add(cell, width)
+    with comp.comb_group(f"{cell}_group") as add_group:
+        add_cell.left = left
+        add_cell.right = right
+    return add_cell, add_group
+
+
+def insert_sub(comp: cb.ComponentBuilder, left, right, cell, width):
+    """Inserts wiring into component {comp} to compute {a} - {b}.
+    1. Within {comp}, creates a combinational group called {cell}_group.
+    2. Within the group, creates a {cell} that computes differences of {width}.
+    3. Puts the values {a} and {b} into {cell}.
+    4. Returns the subtracting cell and the overall group.
+    """
+    sub_cell = comp.sub(cell, width)
+    with comp.comb_group(f"{cell}_group") as sub_group:
+        sub_cell.left = left
+        sub_cell.right = right
+    return sub_cell, sub_group
+
+
+def insert_incr(comp: cb.ComponentBuilder, reg, cell, group):
+    """Inserts wiring into component {comp} to increment {reg} by 1.
+    1. Within component {comp}, creates a group called {group}.
+    2. Within {group}, adds a cell {cell} that computes sums.
+    3. Puts the values of {port} and 1 into {cell}.
+    4. Then puts the answer of the computation back into {port}.
+    4. Returns the group that does this.
+    """
+    incr_cell = comp.add(cell, 32)
+    with comp.group(group) as incr_group:
+        incr_cell.left = reg.out
+        incr_cell.right = 1
+        reg.write_en = 1
+        reg.in_ = incr_cell.out
+        incr_group.done = reg.done
+    return incr_group
+
+
+def insert_decr(comp: cb.ComponentBuilder, reg, cell, group):
+    """Inserts wiring into component {comp} to decrement {reg} by 1.
+    1. Within component {comp}, creates a group called {group}.
+    2. Within {group}, adds a cell {cell} that computes differences.
+    3. Puts the values of {port} and 1 into {cell}.
+    4. Then puts the answer of the computation back into {port}.
+    4. Returns the group that does this.
+    """
+    decr_cell = comp.sub(cell, 32)
+    with comp.group(group) as decr_group:
+        decr_cell.left = reg.out
+        decr_cell.right = cb.const(32, 1)
+        reg.write_en = 1
+        reg.in_ = decr_cell.out
+        decr_group.done = reg.done
+    return decr_group
+
+
+def insert_reg_store(comp: cb.ComponentBuilder, reg, val, group):
+    """Stores a value in a register.
+    1. Within component {comp}, creates a group called {group}.
+    2. Within {group}, sets the register {reg} to {val}.
+    3. Returns the group that does this.
+    """
+    with comp.group(group) as reg_grp:
+        reg.in_ = val
+        reg.write_en = 1
+        reg_grp.done = reg.done
+    return reg_grp
+
+
+def insert_mem_load(comp: cb.ComponentBuilder, mem, i, reg, group):
```

Comment:
This should check if the passed memory is a `std_mem` or not because other the interface of the memory is not correct. In general, I think we want to build a separate `Stdlib` like class for the various different memories we've been defining.